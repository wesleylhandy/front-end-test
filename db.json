{
  "notes": [
    {
      "subject": "Lorem ipsum dolor sit amet",
      "body":
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed non velit vehicula lacus dictum tristique. Quisque ac justo magna. Duis eu est purus. Quisque ut enim tortor. Morbi interdum risus a lorem posuere dapibus. Morbi orci est, varius sed dolor quis, efficitur scelerisque sapien. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Aenean at libero sapien. Morbi ultrices malesuada nisl, sit amet aliquam ipsum egestas non. Nunc at massa vel metus vestibulum fringilla in vitae est. Pellentesque eget lorem eget arcu eleifend pretium nec sed tortor. Cras tristique ex quis aliquam vulputate.",
      "read": true
    },
    {
      "subject": "Aliquam aliquet sem mauris",
      "body":
        "Aliquam aliquet sem mauris, quis blandit sapien faucibus non. Suspendisse potenti. Praesent pulvinar euismod magna, vel consectetur enim pretium nec. Donec in urna scelerisque, lobortis purus quis, porttitor nulla. Ut suscipit gravida nulla in scelerisque. Sed maximus libero massa, non vehicula nisi sollicitudin hendrerit. Mauris lobortis, lectus sit amet rhoncus ultricies, lorem urna euismod mi, et sollicitudin sem tortor sed justo. Fusce massa tellus, fermentum quis quam nec, accumsan ultrices lectus. Integer eu iaculis felis, a consectetur dui. Nulla eleifend nunc justo, ac feugiat turpis finibus at.",
      "read": false
    },
    {
      "subject": "Nam et feugiat mi, ut aliq",
      "body":
        "Nam et feugiat mi, ut aliquet sapien. Vivamus et felis urna. Nullam non bibendum tellus, eget tincidunt purus. Integer eget ultricies ipsum, vel semper purus. Curabitur lorem metus, volutpat sit amet urna vitae, vulputate lobortis elit. Nam lacinia accumsan leo, nec porta tellus mollis non. Praesent nec sodales libero, a consectetur velit. Curabitur luctus nibh erat, quis consequat libero sodales sed. Phasellus condimentum dui sit amet lectus bibendum sollicitudin. Duis quam dui, placerat a placerat ullamcorper, elementum ac nisl. Pellentesque ac dui ac erat aliquet molestie placerat a neque. Praesent sed mollis velit, sit amet egestas massa. Sed malesuada venenatis varius. Vestibulum tincidunt elit vitae pharetra rutrum. Aliquam aliquam tincidunt nulla nec gravida.",
      "read": true
    },
    {
      "subject": "Duis nec libero a sem euis",
      "body":
        "Duis nec libero a sem euismod fringilla at nec nulla. Vestibulum id lectus magna. Nullam venenatis sodales elit, et faucibus neque lobortis ac. In lacinia posuere hendrerit. Etiam vitae nisi sit amet dui imperdiet ornare. Vivamus varius tempor leo, quis lobortis tortor porttitor sed. Pellentesque interdum lectus in sem convallis, quis consectetur ex ultrices. Etiam quis lorem purus. Cras ut vehicula velit. Donec scelerisque felis vel velit tristique pharetra. Aliquam id justo bibendum sapien suscipit ullamcorper. In non felis non nisi posuere accumsan. Aenean sit amet mi erat. Donec id lorem enim. Proin rhoncus nunc sit amet orci imperdiet, in mattis arcu mollis. In hac habitasse platea dictumst.",
      "read": false
    },
    {
      "subject": "Cras tortor lacus, mattis ",
      "body":
        "Cras tortor lacus, mattis et tincidunt eget, posuere in dolor. In eget quam at tellus egestas semper et vel risus. In ut lobortis ex. Nullam congue aliquet odio vel faucibus. In dictum a massa at sagittis. Nunc aliquam tellus et mi laoreet sagittis. Etiam luctus a tortor fringilla facilisis. Nam a nisl efficitur, eleifend magna pellentesque, ultricies odio. Quisque urna tortor, feugiat sagittis interdum id, pellentesque facilisis velit. Vivamus vel gravida purus.",
      "read": false
    }
  ]
}
