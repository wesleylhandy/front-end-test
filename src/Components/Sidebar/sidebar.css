.NotesSidebar {
	background: #a19689;
	border-right: solid 1px #636463;
	color: white;
}

.NotesSidebar__title {
	margin: 0;
	font-size: 1.25rem;
	padding: 0.2rem;
}

.NotesSidebarItem {
	padding: 0 0.3rem;
	display: flex;
	flex-direction: row;
	border-top: solid 1px #636463;
}

.NotesSidebarItem.selected {
	background: #ead592;
}

.NotesSidebarItem__title {
	margin: 0.3rem 0;
	display: inline-block;
	flex: 1;
}
