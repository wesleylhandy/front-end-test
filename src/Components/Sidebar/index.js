import React from "react";
import checkMark from "../../SVG/check-mark.svg";
import classNames from "classnames";
import PropTypes from "prop-types";
import "./sidebar.css";

const Sidebar = ({ notes = [], currentNoteIndex, selectNote }) => (
	<section className="NotesSidebar">
		<h2 className="NotesSidebar__title">Available Notes:</h2>
		<div className="NotesSidebar__list">
			{notes.length ? (
				notes.map(({ subject, read }, idx) => (
					<div
						key={`note-${idx}`}
						className={classNames("NotesSidebarItem", {
							selected: idx === currentNoteIndex,
						})}
						onClick={e => selectNote(idx)}
						id={`note-${idx}`}
						role="button"
					>
						<h4 className="NotesSidebarItem__title">{subject}</h4>
						{read && <img alt="Check Mark" src={checkMark} />}
					</div>
				))
			) : (
				<div className="NotesSidebarItem empty">--Empty--</div>
			)}
		</div>
	</section>
);

Sidebar.propTypes = {
	notes: PropTypes.arrayOf(
		PropTypes.shape({
			subject: PropTypes.string.isRequired,
			body: PropTypes.string.isRequired,
			read: PropTypes.bool.isRequired,
		})
	),
	currentNoteIndex: PropTypes.number.isRequired,
	selectNote: PropTypes.func.isRequired,
};

export default Sidebar;
