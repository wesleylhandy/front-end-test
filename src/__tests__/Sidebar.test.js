import React from "react";
import Sidebar from "../Components/Sidebar";
import { create, act } from "react-test-renderer";
import db from "../../db.json";

let currentNoteIndex = -1;
const notes = [...db.notes];
const selectNote = idx => {currentNoteIndex = idx};
const toggleReadStatus = idx => notes[idx].read = !notes[idx].read

it("renders correctly", () => {
	const tree = create(
		<Sidebar
			notes={notes}
			currentNoteIndex={currentNoteIndex}
			selectNote={selectNote}
		/>
	).toJSON();
	expect(tree).toMatchSnapshot();
});

it("changes currentNodeIndex when a note is clicked", ()=>{
    let component;
	act(() => {
		component = create(
			<Sidebar
                notes={notes}
                currentNoteIndex={currentNoteIndex}
                selectNote={selectNote}
            />
		);
    });
    const buttons = component.root.children[0].children[1].children[0];

    for (let i = 0; i < buttons.length; i++) {
        act(()=>{
            buttons[i].props.onClick();
        })
        expect(currentNoteIndex).toBe(i);
    }
})

it("toggles read check mark when read status is toggled", ()=>{
    let component; let checkMark;
    const randomIdx = Math.floor(Math.random() * notes.length);
    const read = notes[randomIdx].read
	act(() => {
		component = create(
			<Sidebar
                notes={notes}
                currentNoteIndex={currentNoteIndex}
                selectNote={selectNote}
            />
		);
    });
    checkMark = component.root.children[0].children[1].children[randomIdx].children.length === 2
    expect(checkMark).toBe(read);
    expect(component.toJSON()).toMatchSnapshot();
    
    act(()=>{
        toggleReadStatus(randomIdx);
        component.update(<Sidebar
            notes={notes}
            currentNoteIndex={currentNoteIndex}
            selectNote={selectNote}
        />)
    })
    
    checkMark = component.root.children[0].children[1].children[randomIdx].children.length === 2
    expect(checkMark).not.toBe(read);
	expect(component.toJSON()).toMatchSnapshot();
})