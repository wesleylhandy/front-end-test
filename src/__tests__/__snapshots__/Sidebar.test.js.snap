// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="NotesSidebar"
>
  <h2
    className="NotesSidebar__title"
  >
    Available Notes:
  </h2>
  <div
    className="NotesSidebar__list"
  >
    <div
      className="NotesSidebarItem"
      id="note-0"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Lorem ipsum dolor sit amet
      </h4>
      <img
        alt="Check Mark"
        src="check-mark.svg"
      />
    </div>
    <div
      className="NotesSidebarItem"
      id="note-1"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Aliquam aliquet sem mauris
      </h4>
    </div>
    <div
      className="NotesSidebarItem"
      id="note-2"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Nam et feugiat mi, ut aliq
      </h4>
      <img
        alt="Check Mark"
        src="check-mark.svg"
      />
    </div>
    <div
      className="NotesSidebarItem"
      id="note-3"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Duis nec libero a sem euis
      </h4>
    </div>
    <div
      className="NotesSidebarItem"
      id="note-4"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Cras tortor lacus, mattis 
      </h4>
    </div>
  </div>
</section>
`;

exports[`toggles read check mark when read status is toggled 1`] = `
<section
  className="NotesSidebar"
>
  <h2
    className="NotesSidebar__title"
  >
    Available Notes:
  </h2>
  <div
    className="NotesSidebar__list"
  >
    <div
      className="NotesSidebarItem"
      id="note-0"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Lorem ipsum dolor sit amet
      </h4>
      <img
        alt="Check Mark"
        src="check-mark.svg"
      />
    </div>
    <div
      className="NotesSidebarItem"
      id="note-1"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Aliquam aliquet sem mauris
      </h4>
    </div>
    <div
      className="NotesSidebarItem"
      id="note-2"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Nam et feugiat mi, ut aliq
      </h4>
      <img
        alt="Check Mark"
        src="check-mark.svg"
      />
    </div>
    <div
      className="NotesSidebarItem"
      id="note-3"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Duis nec libero a sem euis
      </h4>
    </div>
    <div
      className="NotesSidebarItem"
      id="note-4"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Cras tortor lacus, mattis 
      </h4>
    </div>
  </div>
</section>
`;

exports[`toggles read check mark when read status is toggled 2`] = `
<section
  className="NotesSidebar"
>
  <h2
    className="NotesSidebar__title"
  >
    Available Notes:
  </h2>
  <div
    className="NotesSidebar__list"
  >
    <div
      className="NotesSidebarItem"
      id="note-0"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Lorem ipsum dolor sit amet
      </h4>
      <img
        alt="Check Mark"
        src="check-mark.svg"
      />
    </div>
    <div
      className="NotesSidebarItem"
      id="note-1"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Aliquam aliquet sem mauris
      </h4>
    </div>
    <div
      className="NotesSidebarItem"
      id="note-2"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Nam et feugiat mi, ut aliq
      </h4>
      <img
        alt="Check Mark"
        src="check-mark.svg"
      />
    </div>
    <div
      className="NotesSidebarItem"
      id="note-3"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Duis nec libero a sem euis
      </h4>
      <img
        alt="Check Mark"
        src="check-mark.svg"
      />
    </div>
    <div
      className="NotesSidebarItem"
      id="note-4"
      onClick={[Function]}
      role="button"
    >
      <h4
        className="NotesSidebarItem__title"
      >
        Cras tortor lacus, mattis 
      </h4>
    </div>
  </div>
</section>
`;
